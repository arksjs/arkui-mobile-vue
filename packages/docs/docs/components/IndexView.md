# IndexView/IndexViewItem 索引栏

<CodeDemo name="IndexView">

<<< @/../../demo/src/components/Navigation/IndexView/index.vue

</CodeDemo>

## Import

```js
import { TaIndexView, TaIndexViewItem } from 'tantalum-ui-mobile'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type { IndexViewOnChange, IndexViewRef } from 'tantalum-ui-mobile'
```

## IndexView Props

| 属性              | 类型             | 默认值 | 必填 | 说明                        |
| ----------------- | ---------------- | ------ | ---- | --------------------------- |
| sticky-offset-top | string \| number | 0      | 否   | 数值默认是 px，也支持 vw/vh |

## IndexView Events

| 事件   | 描述       | 回调函数参数                        | TypeScript 函数   |
| ------ | ---------- | ----------------------------------- | ----------------- |
| change | 切换时触发 | (name: string, activeIndex: number) | IndexViewOnChange |

## Methods

```ts
interface IndexViewRef {
  scrollTo: (name: string) => void
  scrollToIndex: (index: number) => void
}
```

| 方法名        | 说明                     |
| ------------- | ------------------------ |
| scrollTo      | 切换到指定 name 的 Item  |
| scrollToIndex | 切换到指定 index 的 Item |

## IndexView Slots

### #default

注：其中只可放置 [IndexViewItem](./IndexView.md#indexviewitem-索引子项) 组件，否则会导致未定义的行为。

```vue
<ta-index-view>
  <ta-index-view-item name="A">
    <ta-cell label="单元格"></ta-cell>
    <ta-cell label="单元格"></ta-cell>
    <ta-cell label="单元格"></ta-cell>
  </ta-index-view-item>
  ...
</ta-index-view>
```

## IndexViewItem Props

| 属性  | 类型   | 默认值 | 必填 | 说明                                                               |
| ----- | ------ | ------ | ---- | ------------------------------------------------------------------ |
| name  | string |        | 是   | 唯一标识，设置后配合 ScrollTab 组件的 `v-model` 和 `onChange` 使用 |
| title | string |        | 否   | 分组名，也应用于吸附侧边栏，如果没有设置则获取 `name` 的值         |

## IndexViewItem Slots

### #default

```vue
<ta-index-view-item name="A">
  <ta-cell label="单元格"></ta-cell>
  <ta-cell label="单元格"></ta-cell>
  <ta-cell label="单元格"></ta-cell>
  ...
</ta-index-view-item>
```
