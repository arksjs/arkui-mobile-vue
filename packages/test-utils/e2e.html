<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
