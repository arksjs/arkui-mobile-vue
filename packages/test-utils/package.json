{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
