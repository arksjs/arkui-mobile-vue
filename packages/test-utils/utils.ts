export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
