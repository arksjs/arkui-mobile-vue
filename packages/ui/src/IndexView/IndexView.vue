<template>
  <div class="ak-index-view">
    <div class="ak-index-view_sidebar">
      <ul class="ak-index-view_list" ref="navEl">
        <li
          :class="{ active: item.value === activeName }"
          v-for="(item, index) in indexList"
          :data-value="item.value"
          :data-index="index"
          :key="item.value"
        >
          {{ item.label }}
        </li>
      </ul>
    </div>
    <div class="ak-index-view_body">
      <StickyView
        :offsetTop="stickyOffsetTop"
        :modelValue="modelValue"
        ref="bodyRef"
        @resetItems="onResetItems"
        @change="onChange"
      >
        <slot></slot>
      </StickyView>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref, watch } from 'vue'
import { StickyView } from '../StickyView'
import { sizeValidator } from '../helpers/validator'
import { isString, rangeInteger } from '../helpers/util'
import type {
  OnResetItems,
  StickyViewRef,
  OnChange as StickyViewOnChange
} from '../StickyView/types'
import { useTouch } from '../hooks/use-touch'
import { emitChangeValidator } from '../StickyView/props'
import type { ResetContainer } from '../Sticky/types'
import type { PropsToEmits } from '../helpers/types'
import type { IndexViewEmits } from './types'

export default defineComponent({
  name: 'ak-index-view',
  components: { StickyView },
  props: {
    modelValue: {
      type: String
    },
    stickyOffsetTop: {
      validator: sizeValidator,
      default: 0
    }
  },
  emits: {
    'update:modelValue': name => isString(name),
    change: emitChangeValidator
  } as PropsToEmits<IndexViewEmits>,
  setup(props, { emit, expose }) {
    const navEl = ref<HTMLElement>()
    const bodyRef = ref<StickyViewRef>()
    const indexList = ref<
      {
        value: string
        label: string
      }[]
    >([])
    const activeName = ref<string>()

    // 单独更新以下tab的activeName
    function updateActiveName(name?: string) {
      if (name != null && isInTab(name) && name !== activeName.value) {
        activeName.value = name
      }
    }

    function isInTab(name: string) {
      for (let i = 0; i < indexList.value.length; i++) {
        if (indexList.value[i].value === name) {
          return true
        }
      }

      return false
    }

    const resetContainer: ResetContainer = containSelector => {
      bodyRef.value?.resetContainer(containSelector)
    }

    const onResetItems: OnResetItems = items => {
      indexList.value = items.map(item => {
        return {
          value: item.name,
          label: item.title
        }
      })
    }

    const onChange: StickyViewOnChange = (name, index) => {
      updateActiveName(name)
      emit('update:modelValue', name)
      emit('change', name, index)
    }

    /**
     * 滚动到第index个
     */
    function scrollToIndex(index: number) {
      bodyRef.value?.scrollToIndex(index)
    }

    /**
     * 滚动到指定name
     */
    function scrollTo(name: string) {
      bodyRef.value?.scrollTo(name)
    }

    let coords: {
      size: number
      offsetY: number
      startY: number
      current: number
      isChange: boolean
    } | null = null
    let changeTimer: number

    useTouch({
      el: navEl,
      onTouchStart(e) {
        const { clientY } = e.touchObject

        const $target = e.target as HTMLElement
        const index = parseInt($target.dataset.index as string)
        const rects = $target.getClientRects()[0]

        coords = {
          size: rects.height,
          offsetY: clientY - rects.top,
          startY: clientY,
          current: index,
          isChange: false
        }

        clearTimeout(changeTimer)
        changeTimer = window.setTimeout(() => {
          scrollToIndex(index)
        }, 500)

        e.preventDefault()
      },

      onTouchMove(e) {
        if (!coords) {
          return
        }

        const { clientY } = e.touchObject
        const { startY, size, offsetY, current } = coords

        const y = clientY - startY

        let offsetCount = 0

        if (y > 0) {
          offsetCount =
            Math.floor(y / size) + (y % size > size - offsetY ? 1 : 0)
        } else if (y < 0) {
          offsetCount = -Math.floor(-y / size) + (-y % size > offsetY ? -1 : 0)
        }

        if (offsetCount !== 0) {
          clearTimeout(changeTimer)
          coords.isChange = true

          changeTimer = window.setTimeout(() => {
            scrollToIndex(
              rangeInteger(current + offsetCount, 0, indexList.value.length - 1)
            )
          }, 100)
        }

        e.stopPropagation()
      },

      onTouchEnd(e) {
        if (coords) {
          if (!coords.isChange) {
            clearTimeout(changeTimer)
            scrollToIndex(coords.current)
          }

          coords = null
          e.stopPropagation()
        }
      }
    })

    watch(
      () => props.modelValue,
      val => updateActiveName(val)
    )

    onMounted(() => {
      resetContainer(document.documentElement)
      updateActiveName(props.modelValue)
      if (activeName.value == null && indexList.value.length > 0) {
        // 首次要写入一个
        activeName.value = indexList.value[0].value
      }
    })

    expose({
      scrollTo,
      scrollToIndex,
      resetContainer
    })

    return {
      navEl,
      bodyRef,
      activeName,
      indexList,
      onChange,
      onResetItems
    }
  }
})
</script>
