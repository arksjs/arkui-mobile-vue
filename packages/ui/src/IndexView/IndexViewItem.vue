<template>
  <Group
    class="fx-sticky-view-item fx-index-view-item"
    :title="name"
    :data-name="name"
  >
    <slot></slot>
  </Group>
</template>

<script lang="ts">
import { defineComponent, inject, onMounted, onUnmounted } from 'vue'
import { Group } from '../Group'
import { createUpdateInItem } from '../hooks/use-list'

export default defineComponent({
  name: 'fx-index-view-item',
  components: { Group },
  props: {
    name: {
      type: String,
      required: true
    }
  },
  setup() {
    const update = inject(
      'fxStickyViewUpdate',
      createUpdateInItem('index-view')
    )

    onMounted(update)
    onUnmounted(update)

    return {}
  }
})
</script>
