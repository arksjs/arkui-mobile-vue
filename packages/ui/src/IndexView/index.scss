@import '../style/var.scss';

.ak-index-view {
  overflow-y: auto;
  position: relative;

  &_sidebar {
    position: fixed;
    right: 0;
    top: 50%;
    transform: translate3d(0, -50%, 0);
    width: 36px;
    z-index: 1000;
    display: flex;
    flex-direction: column;
    justify-content: center;
    overflow: hidden;
  }

  &_list {
    list-style: none;
    margin: 0;
    padding: 0;
    user-select: none;

    li {
      padding: 0 0 0 10px;
      text-align: center;
      height: 18px;
      line-height: 18px;
      font-size: 10px;
      color: $font3-color;
      font-weight: 500;
      cursor: pointer;

      &.active {
        color: $primary-color;
      }
    }
  }
}

.ak-index-view-item {
  margin-bottom: 0;

  .ak-group_header {
    height: 28px;
    background-color: $background-color;
  }
}
