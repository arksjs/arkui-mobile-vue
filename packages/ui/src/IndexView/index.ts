import { multiWithInstall } from '../helpers/with-install'
import IndexView from './IndexView.vue'
import IndexViewItem from './IndexViewItem.vue'

export { IndexView, IndexViewItem }
export default IndexView
export const install = multiWithInstall(IndexView, [IndexViewItem])
