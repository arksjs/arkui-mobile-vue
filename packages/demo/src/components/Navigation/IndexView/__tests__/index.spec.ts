import { mount } from '@vue/test-utils'
import { IndexView, IndexViewItem } from '@/IndexView'

describe('IndexView', () => {
  test('IndexView should render default correctly', () => {
    const wrapper = mount(IndexView, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('IndexViewItem should render default correctly', () => {
    const wrapper = mount(IndexViewItem, {
      props: { name: 'item-1' },
      slots: {
        default: 'content'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
