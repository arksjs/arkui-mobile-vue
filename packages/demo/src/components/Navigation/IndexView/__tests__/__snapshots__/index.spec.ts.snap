// Vitest Snapshot v1

exports[`IndexView > IndexView should render default correctly 1`] = `
"<div class=\\"fx-index-view\\">
  <div class=\\"fx-index-view_sidebar\\">
    <ul class=\\"fx-index-view_list\\"></ul>
  </div>
  <div class=\\"fx-index-view_body\\">
    <div class=\\"fx-sticky-view\\">
      <div class=\\"fx-sticky-view_list\\"></div>
      <div class=\\"fx-sticky fx-sticky-view_top\\">
        <div class=\\"fx-sticky_content\\">
          <div class=\\"fx-sticky-view_fixed\\">
            <div class=\\"fx-sticky-view_fixed-inner\\" style=\\"transform: translate3d(0px, -100%, 0px);\\"></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`IndexView > IndexViewItem should render default correctly 1`] = `
"<div class=\\"fx-group fx-horizontal-hairline fx-sticky-view-item fx-index-view-item\\" data-name=\\"item-1\\">
  <div class=\\"fx-group_header\\">
    <div class=\\"fx-group_title\\">item-1</div>
    <div class=\\"fx-group_more\\"></div>
  </div>
  <div class=\\"fx-group_body fx-horizontal-hairline hairline-reverse\\">
    <div class=\\"fx-group_body-inner\\">content</div>
  </div>
</div>"
`;
